### PR TITLE
feat: add stringify function

### PR DESCRIPTION
**Pull request by magicmatatjahu**
**Description**

Add static `stringify` and `parse` functions to the `AsyncAPIDocument` class"
- handle circular references in the `stringify` function, `JSON.parse` isn't compatible with `AsyncAPIDocument.parse`
- add unit tests

More info why need that function -> https://github.com/asyncapi/parser-js/issues/266

The algorithm of stringify works in this way that it finds all the same references and for all of them (except the first one encountered) converts it to the correct path using a JSON Pointer. Why? For two reasons:
- to support circular references
- to make the output string as small as possible

Example with circular references:

<details>
<summary><b>Example with circular references:</b></summary>

Any folded content here. It requires an empty line just above it.

```yaml
asyncapi: 2.0.0
info:
  title: My Circular API
  version: '1.0.0'
channels:
  recursive:
    subscribe:
      message:
        payload:
          $ref: '#/components/schemas/RecursiveSelf'
  external/file:
    publish:
      message:
        payload:
          $ref: './good/circular-refs-file-ref.yaml#/ExternalFile'
  nonRecursive:
    subscribe:
      message:
        payload:
          $ref: '#/components/schemas/NonRecursive'
  testChannel:
    subscribe:
      message:
        oneOf:
          - $ref: '#/components/messages/testMessage'
components:
  messages:
    testMessage:
      contentType: application/json
      payload:
        $ref: '#/components/schemas/NormalSchemaA'
  schemas:
    NonRecursive:
      type: object
      properties:
        child:
          $ref: '#/components/schemas/NonRecursiveChild'
    NonRecursiveChild:
      type: object
      properties:
        value:
          type: string
    RecursiveSelf:
      type: object
      properties:
        selfChildren:
          type: array
          items:
            $ref: '#/components/schemas/RecursiveSelf'
        selfObjectChildren:
          type: object
          properties:
            test:
              $ref: '#/components/schemas/RecursiveSelf'
            nonRecursive:
              type: string
        selfSomething:
          type: object
          properties:
            test: 
               $ref: '#/components/schemas/RecursiveAncestor'
    RecursiveAncestor:
      type: object
      properties:
        ancestorChildren:
          type: array
          items:
            $ref: '#/components/schemas/RecursiveSelf'
        ancestorSomething:
          type: string
    NormalSchemaA:
      type: object
      properties:
        schemaBReference:
          $ref: '#/components/schemas/NormalSchemaB'
        schemaCReference:
          $ref: '#/components/schemas/NormalSchemaC'
        commonEnumName:
          type: string
          enum:
            - ENUM_1
            - ENUM_2
    NormalSchemaB:
      type: string
      enum:
        - ENUM_A
        - ENUM_B
        - ENUM_C
        - ENUM_D
    NormalSchemaC:
      allOf:
        - $ref: '#/components/schemas/NormalSchemaB'
        - type: string
          enum:
            - ENUM_E
    NestedAllOfSchema:
      allOf:
        - $ref: '#/components/schemas/NormalSchemaA'
        - type: object
          properties:
            parent:
              allOf:
                - $ref: '#/components/schemas/NestedAllOfSchema'
                - $ref: '#/components/schemas/NormalSchemaA'
            name:
              type: string
          required:
            - name
    OneOf:
      type: object
      properties: 
        kind: 
          oneOf:
            - $ref: '#/components/schemas/OneOf'
            - type: string
            - enum:
              - boolean
              - string 
    AnyOf:
      anyOf:
        - type: integer
        - type: number
        - type: string
        - type: boolean
        - type: object
        - type: array
          items:
            $ref: "#/components/schemas/AnyOf"
    RecursiveComplex:
        type: [object, array]
        patternProperties:
          ^foo: 
            $ref: '#/components/schemas/RecursiveSelf'
          ^bar: 
            type: string
        contains:
          $ref: '#/components/schemas/RecursiveComplex'
        items:
          - type: string
          - $ref: '#/components/schemas/RecursiveComplex'
        if:
          $ref: '#/components/schemas/RecursiveAncestor'
        then:
          $ref: '#/components/schemas/RecursiveComplex'
```

</details>

<details>
<summary><b>Stringified</b></summary>

```json
{
  "asyncapi": "2.0.0",
  "info": {
    "title": "My Circular API",
    "version": "1.0.0"
  },
  "channels": {
    "recursive": {
      "subscribe": {
        "message": {
          "payload": {
            "type": "object",
            "properties": {
              "selfChildren": {
                "type": "array",
                "items": "$ref:$.channels.recursive.subscribe.message.payload",
                "x-parser-schema-id": "<anonymous-schema-1>"
              },
              "selfObjectChildren": {
                "type": "object",
                "properties": {
                  "test": "$ref:$.channels.recursive.subscribe.message.payload",
                  "nonRecursive": {
                    "type": "string",
                    "x-parser-schema-id": "<anonymous-schema-3>"
                  }
                },
                "x-parser-schema-id": "<anonymous-schema-2>"
              },
              "selfSomething": {
                "type": "object",
                "properties": {
                  "test": {
                    "type": "object",
                    "properties": {
                      "ancestorChildren": {
                        "type": "array",
                        "items": "$ref:$.channels.recursive.subscribe.message.payload",
                        "x-parser-schema-id": "<anonymous-schema-5>"
                      },
                      "ancestorSomething": {
                        "type": "string",
                        "x-parser-schema-id": "<anonymous-schema-6>"
                      }
                    },
                    "x-parser-schema-id": "RecursiveAncestor"
                  }
                },
                "x-parser-schema-id": "<anonymous-schema-4>"
              }
            },
            "x-parser-schema-id": "RecursiveSelf"
          },
          "x-parser-original-schema-format": "application/vnd.aai.asyncapi;version=2.0.0",
          "x-parser-original-payload": "$ref:$.channels.recursive.subscribe.message.payload",
          "schemaFormat": "application/vnd.aai.asyncapi;version=2.0.0",
          "x-parser-message-parsed": true,
          "x-parser-message-name": "<anonymous-message-1>"
        }
      }
    },
    "external/file": {
      "publish": {
        "message": {
          "payload": {
            "type": "object",
            "properties": {
              "testExt": {
                "type": "object",
                "properties": {
                  "children": {
                    "type": "array",
                    "items": "$ref:$.channels.external/file.publish.message.payload",
                    "x-parser-schema-id": "<anonymous-schema-9>"
                  }
                },
                "x-parser-schema-id": "<anonymous-schema-8>"
              }
            },
            "x-parser-schema-id": "<anonymous-schema-7>"
          },
          "x-parser-original-schema-format": "application/vnd.aai.asyncapi;version=2.0.0",
          "x-parser-original-payload": "$ref:$.channels.external/file.publish.message.payload",
          "schemaFormat": "application/vnd.aai.asyncapi;version=2.0.0",
          "x-parser-message-parsed": true,
          "x-parser-message-name": "<anonymous-message-2>"
        }
      }
    },
    "nonRecursive": {
      "subscribe": {
        "message": {
          "payload": {
            "type": "object",
            "properties": {
              "child": {
                "type": "object",
                "properties": {
                  "value": {
                    "type": "string",
                    "x-parser-schema-id": "<anonymous-schema-10>"
                  }
                },
                "x-parser-schema-id": "NonRecursiveChild"
              }
            },
            "x-parser-schema-id": "NonRecursive"
          },
          "x-parser-original-schema-format": "application/vnd.aai.asyncapi;version=2.0.0",
          "x-parser-original-payload": {
            "type": "object",
            "properties": {
              "child": {
                "type": "object",
                "properties": {
                  "value": {
                    "type": "string"
                  }
                }
              }
            }
          },
          "schemaFormat": "application/vnd.aai.asyncapi;version=2.0.0",
          "x-parser-message-parsed": true,
          "x-parser-message-name": "<anonymous-message-3>"
        }
      }
    },
    "testChannel": {
      "subscribe": {
        "message": {
          "oneOf": [
            {
              "contentType": "application/json",
              "payload": {
                "type": "object",
                "properties": {
                  "schemaBReference": {
                    "type": "string",
                    "enum": [
                      "ENUM_A",
                      "ENUM_B",
                      "ENUM_C",
                      "ENUM_D"
                    ],
                    "x-parser-schema-id": "NormalSchemaB"
                  },
                  "schemaCReference": {
                    "allOf": [
                      "$ref:$.channels.testChannel.subscribe.message.oneOf[0].payload.properties.schemaBReference",
                      {
                        "type": "string",
                        "enum": [
                          "ENUM_E"
                        ],
                        "x-parser-schema-id": "<anonymous-schema-11>"
                      }
                    ],
                    "x-parser-schema-id": "NormalSchemaC"
                  },
                  "commonEnumName": {
                    "type": "string",
                    "enum": [
                      "ENUM_1",
                      "ENUM_2"
                    ],
                    "x-parser-schema-id": "<anonymous-schema-12>"
                  }
                },
                "x-parser-schema-id": "NormalSchemaA"
              },
              "x-parser-original-schema-format": "application/vnd.aai.asyncapi;version=2.0.0",
              "x-parser-original-payload": {
                "type": "object",
                "properties": {
                  "schemaBReference": {
                    "type": "string",
                    "enum": [
                      "ENUM_A",
                      "ENUM_B",
                      "ENUM_C",
                      "ENUM_D"
                    ]
                  },
                  "schemaCReference": {
                    "allOf": [
                      "$ref:$.channels.testChannel.subscribe.message.oneOf[0].x-parser-original-payload.properties.schemaBReference",
                      {
                        "type": "string",
                        "enum": [
                          "ENUM_E"
                        ]
                      }
                    ]
                  },
                  "commonEnumName": {
                    "type": "string",
                    "enum": [
                      "ENUM_1",
                      "ENUM_2"
                    ]
                  }
                }
              },
              "schemaFormat": "application/vnd.aai.asyncapi;version=2.0.0",
              "x-parser-message-parsed": true,
              "x-parser-message-name": "testMessage"
            }
          ]
        }
      }
    }
  },
  "components": {
    "messages": {
      "testMessage": "$ref:$.channels.testChannel.subscribe.message.oneOf[0]"
    },
    "schemas": {
      "NonRecursive": "$ref:$.channels.nonRecursive.subscribe.message.payload",
      "NonRecursiveChild": "$ref:$.channels.nonRecursive.subscribe.message.payload.properties.child",
      "RecursiveSelf": "$ref:$.channels.recursive.subscribe.message.payload",
      "RecursiveAncestor": "$ref:$.channels.recursive.subscribe.message.payload.properties.selfSomething.properties.test",
      "NormalSchemaA": "$ref:$.channels.testChannel.subscribe.message.oneOf[0].payload",
      "NormalSchemaB": "$ref:$.channels.testChannel.subscribe.message.oneOf[0].payload.properties.schemaBReference",
      "NormalSchemaC": "$ref:$.channels.testChannel.subscribe.message.oneOf[0].payload.properties.schemaCReference",
      "NestedAllOfSchema": {
        "allOf": [
          "$ref:$.channels.testChannel.subscribe.message.oneOf[0].payload",
          {
            "type": "object",
            "properties": {
              "parent": {
                "allOf": [
                  "$ref:$.components.schemas.NestedAllOfSchema",
                  "$ref:$.channels.testChannel.subscribe.message.oneOf[0].payload"
                ],
                "x-parser-schema-id": "<anonymous-schema-14>"
              },
              "name": {
                "type": "string",
                "x-parser-schema-id": "<anonymous-schema-15>"
              }
            },
            "required": [
              "name"
            ],
            "x-parser-schema-id": "<anonymous-schema-13>"
          }
        ],
        "x-parser-schema-id": "NestedAllOfSchema"
      },
      "OneOf": {
        "type": "object",
        "properties": {
          "kind": {
            "oneOf": [
              "$ref:$.components.schemas.OneOf",
              {
                "type": "string",
                "x-parser-schema-id": "<anonymous-schema-17>"
              },
              {
                "enum": [
                  "boolean",
                  "string"
                ],
                "x-parser-schema-id": "<anonymous-schema-18>"
              }
            ],
            "x-parser-schema-id": "<anonymous-schema-16>"
          }
        },
        "x-parser-schema-id": "OneOf"
      },
      "AnyOf": {
        "anyOf": [
          {
            "type": "integer",
            "x-parser-schema-id": "<anonymous-schema-19>"
          },
          {
            "type": "number",
            "x-parser-schema-id": "<anonymous-schema-20>"
          },
          {
            "type": "string",
            "x-parser-schema-id": "<anonymous-schema-21>"
          },
          {
            "type": "boolean",
            "x-parser-schema-id": "<anonymous-schema-22>"
          },
          {
            "type": "object",
            "x-parser-schema-id": "<anonymous-schema-23>"
          },
          {
            "type": "array",
            "items": "$ref:$.components.schemas.AnyOf",
            "x-parser-schema-id": "<anonymous-schema-24>"
          }
        ],
        "x-parser-schema-id": "AnyOf"
      },
      "RecursiveComplex": {
        "type": [
          "object",
          "array"
        ],
        "patternProperties": {
          "^foo": "$ref:$.channels.recursive.subscribe.message.payload",
          "^bar": {
            "type": "string",
            "x-parser-schema-id": "<anonymous-schema-25>"
          }
        },
        "contains": "$ref:$.components.schemas.RecursiveComplex",
        "items": [
          {
            "type": "string",
            "x-parser-schema-id": "<anonymous-schema-26>"
          },
          "$ref:$.components.schemas.RecursiveComplex"
        ],
        "if": "$ref:$.channels.recursive.subscribe.message.payload.properties.selfSomething.properties.test",
        "then": "$ref:$.components.schemas.RecursiveComplex",
        "x-parser-schema-id": "RecursiveComplex"
      }
    }
  },
  "x-parser-circular": true,
  "x-parser-spec-parsed": true
}
```

</details>

@jonaslagoni @smoya You should be interest :)

@jonaslagoni I remember that you wanna have such a feature for every models, not only form AsyncAPIDocument model, but there is problem that static methods (also static properties) cannot be inherited, so we have to update all models to support that. Atm I think that stringifing and parsing only main model is enough, if someone will raise feature for stringifing also other models, we can think about that, but not now. 

**Related issue(s)**
Resolves https://github.com/asyncapi/parser-js/issues/266

### CODE REVIEW

**Comment by jonaslagoni** [comment]
@magicmatatjahu can you ping us once the blocked issue is resolved. No need to double review code 😄

**Comment by magicmatatjahu** [comment]
@smoya @jonaslagoni PR is ready to review.

File: lib/models/asyncapi.js [hunk 1]
```diff
@@ -234,6 +234,103 @@ class AsyncAPIDocument extends Base {
...
+        parent[String(field)] = parsedJSON;
+      }
+  
+      // traverse all keys, only if object is array/object
+      if (objOrPath === Object(objOrPath)) {
+        for (const f in objOrPath) {
+          traverse(objOrPath, f);
+        }
+      }
+    }
```
Comment (jonaslagoni) [inline lib/models/asyncapi.js:290]: What is the reason behind this function? 🤔

File: test/models/asyncapi_test.js [hunk 2]
```diff
@@ -978,6 +984,88 @@ describe('AsyncAPIDocument', function() {
...
+
+    it('should parse stringified document with circular references', async function() {
+      // Test circular references to ensure that every circular reference has this same reference after parsing
+      const result = AsyncAPIDocument.parse(circularOutputYAML);
+
+      // not testing on a model level as required xParserCircle value is added before model construction so we need to test through calling parser function
+      expect(result.hasCircular()).to.equal(true);
+
+      // we want false here, even though this schema has some circular refs in some props, it is not circular, but just specific items
+      expect(result.components().schema('RecursiveSelf').isCircular()).to.equal(false);
```
Comment (jonaslagoni) [inline test/models/asyncapi_test.js:1016]: I find way of reusing variables easier to read in terms of the expected tests:
```suggestion
      const resultComponents = result.components();
      const recursiveSelfComponent = resultComponents.schema('RecursiveSelf');
      expect(recursiveSelfComponent.isCircular()).to.equal(false);
```

If it is just one or two recurring times it is used I don't mind, but when it's 2+ times I find this easier to read and to maintain. What do you think?

This needs to be applied below as well of course 🙂

File: test/parse_test.js [hunk 3]
```diff
@@ -901,7 +901,7 @@ describe('memory usage', function () {
     this.timeout(12500);
     const asyncapi = fs.readFileSync(path.resolve(__dirname, './good/zbos_mqtt-all-asyncapi.json'), 'utf8');
 
-    for (let i = 0, l = 50; i < l; i++) {
+    for (let i = 0, l = 25; i < l; i++) {
```
Comment (jonaslagoni) [inline test/parse_test.js:904]: Why this change? 🤔

File: test/models/asyncapi_test.js [hunk 4]
```diff
@@ -978,6 +984,88 @@ describe('AsyncAPIDocument', function() {
   });
   /* eslint-enable sonarjs/cognitive-complexity */
 
+  describe('#stringify()', function() {
+    it('should stringify simple document', async function() {
+      const doc = await parser.parse(simpleInputJSON, { path: path.join(__filename, '../../') });
+      const stringified = AsyncAPIDocument.stringify(doc);
+      expect(stringified).to.be.equal(simpleOutputJSON);
```
Comment (jonaslagoni) [inline test/models/asyncapi_test.js:991]: I wished we used snapshots here, trying to look at the output JSON in a oneliner is 😅  😆

File: lib/models/asyncapi.js [hunk 5]
```diff
@@ -234,6 +234,103 @@ class AsyncAPIDocument extends Base {
...
+  /* eslint-enabled sonarjs/cognitive-complexity */
+}
+
+/**
+ * Replacer function (that transforms the result) for AsyncAPI.stringify() function.
+ * Handles circular references by replacing it by JSONPath notation. 
+ * 
+ * @private
+ */
+function refReplacer() {
```
Comment (jonaslagoni) [inline lib/models/asyncapi.js:304]: Interesting 🤔 

So instead of just printing `Circular` you actually print the reference to the location of the circular schema? 🤯

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
Left a few comments 😄

File: lib/models/asyncapi.js (hunk 5 above)
Comment (magicmatatjahu) [reply lib/models/asyncapi.js:304]: I cannot print `Circular` (or annotate object as circular) due to fact that I need the whole object after parsing :) There is a also deeper problem, because if in one place given object is circular, in another it don't be, so I cannot also change the references to this object from POV of it parent 😄  Yeah, it's very hard to understand, if you don't, please write, I will write example.

File: test/models/asyncapi_test.js (hunk 4 above)
Comment (magicmatatjahu) [reply test/models/asyncapi_test.js:991]: We do not use any snapshots in the unit tests of functions. I know this can be tiresome, but I am consistent with the other tests in this repository.

Comment (jonaslagoni) [reply test/models/asyncapi_test.js:991]: I know, it was only a comment not a suggestion to change 😄

File: test/models/asyncapi_test.js (hunk 2 above)
Comment (magicmatatjahu) [reply test/models/asyncapi_test.js:1016]: On the one hand ok, but on the other hand not, because if I make a variable for each object, I will have to remember that the variable points to the element, and now I have a full track of what compares to what. But I will stay with the current form..

File: lib/models/asyncapi.js (hunk 1 above)
Comment (magicmatatjahu) [reply lib/models/asyncapi.js:290]: If we call the `stringify` function then later each circular reference is stored as a path to the object (we use JSON Pointer for this) - check example in the PR's description. This function just parses the whole string and looks for the given paths and turns them into reference :) This can be compared to the `json-schema-ref-parser` library, but much simpler and only looks for references in the given object and not outside :)

File: test/parse_test.js (hunk 3 above)
Comment (magicmatatjahu) [reply test/parse_test.js:904]: In Linux and MacOS everything works but I realised that sometimes windows has timeout on this test, I don't want to increase timeout (I don't want wait 20s) so I decided to decrease iterations.

File: lib/models/asyncapi.js (hunk 1 above)
Comment (jonaslagoni) [reply lib/models/asyncapi.js:308]: Got it 👍

File: lib/models/asyncapi.js [hunk 6]
```diff
@@ -234,6 +234,121 @@ class AsyncAPIDocument extends Base {
...
+    const pathToObj = new Map();
+    
+    function traverse(parent, field) {
+      let objOrPath = parent;
+      let path = '$ref:$';
+  
+      if (field !== undefined) {
+        // here can be string with `$ref` prefix or normal value 
+        objOrPath = parent[String(field)];
+        const concatenatedPath = field ? `.${  field}` : '';
```
Comment (jonaslagoni) [inline lib/models/asyncapi.js:287]:
```suggestion
        const concatenatedPath = field ? `.${field}` : '';
```

File: lib/models/asyncapi.js [hunk 7]
```diff
@@ -234,6 +234,121 @@ class AsyncAPIDocument extends Base {
...
+        for (const f in objOrPath) {
+          traverse(objOrPath, f);
+        }
+      }
+    }
+  
+    traverse(parsedJSON);
+    return new AsyncAPIDocument(parsedJSON);
+  }
+  /* eslint-enabled sonarjs/cognitive-complexity */
```
Comment (jonaslagoni) [inline lib/models/asyncapi.js:313]:
```suggestion
```

File: lib/models/asyncapi.js [hunk 8]
```diff
@@ -234,6 +234,121 @@ class AsyncAPIDocument extends Base {
...
+   * @returns {string}
+   */
+  static stringify(doc, space) {
+    const rawDoc = doc.json();
+    const copiedDoc = { ...rawDoc };
+    copiedDoc[String(xParserSpecStringified)] = true;
+    return JSON.stringify(copiedDoc, refReplacer(), space);
+  }
+
+  /* eslint-disable sonarjs/cognitive-complexity */
```
Comment (jonaslagoni) [inline lib/models/asyncapi.js:253]: Instead of disabling sonar completely, would it not be an idea to have all this in a separate support file called `stringify.js` 🤔

Would make it much cleaner in terms of implementation as you can avoid nested functions. Also easier to test.

File: lib/models/asyncapi.js [hunk 9]
```diff
@@ -234,6 +234,121 @@ class AsyncAPIDocument extends Base {
   traverseSchemas(callback, schemaTypesToIterate) {
     traverseAsyncApiDocument(this, callback, schemaTypesToIterate);
   }
+
+  /**
+   * Converts a valid AsyncAPI document to a JavaScript Object Notation (JSON) string.
+   * A stringified AsyncAPI document using this function should be parsed via the AsyncAPIDocument.parse() function - the JSON.parse() function is not compatible.
```
Comment (jonaslagoni) [inline lib/models/asyncapi.js:240]: Do you think we maybe should provide a more in-depth explanation in terms of why it is needed? 🤔

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
Have some small suggestions 🙂 

Overall I think we need some better documentation in the main docs, instead of only in  JSDoc. What do you think?

File: lib/models/asyncapi.js (hunk 8 above)
Comment (magicmatatjahu) [reply lib/models/asyncapi.js:253]: I was thinking about it, but those functions are currently used only in AsyncAPIDocument class and I don't know if it makes sense to extract them. I also need nested functions to access variables defined above. I can always make them inline, but then I will have to pass those variables as arguments - is it worth it? 😄

File: lib/models/asyncapi.js (hunk 9 above)
Comment (magicmatatjahu) [reply lib/models/asyncapi.js:240]: Ok, I will describe that functionality in Readme.md :)

File: lib/models/asyncapi.js (hunk 8 above)
Comment (jonaslagoni) [reply lib/models/asyncapi.js:253]: I mean, it just makes the functions complex to look at, as long as you tried to imagine how it would be for someone who has never seen the code, then all good 🙂 Your call.

Comment (magicmatatjahu) [reply lib/models/asyncapi.js:253]: As inline they will still be complex 😆  Maybe a better solution would be to increase the complex function standard from 15 to 20 or 25? In many libraries we end up at around 17/18 and then this would fix the problem.

Comment (jonaslagoni) [reply lib/models/asyncapi.js:253]: The underlying problem is not sonar triggering cognitive-complexity 😄 You could easily refactor it to use smaller external functions and provide the missing variables as arguments, but it requires some restructuring of course.

Comment (magicmatatjahu) [reply lib/models/asyncapi.js]: Fixed :P

**Comment by magicmatatjahu** [comment]
@jonaslagoni Ok, I updated Readme and also fixed complexity of functions :) Could you check again?

**Review by jonaslagoni — DISMISSED** [review]
LGTM 👍

**Comment by magicmatatjahu** [comment]
@jonaslagoni Thanks!

@alequetzalli Do you wanna make review of Readme.md's changes? :)

File: README.md [hunk 10]
```diff
@@ -218,6 +218,16 @@ Parser dereferences all circular references by default. In addition, to simplify
...
 
+## Stringify
+
+Converting a parsed document to a string may be necessary when saving the parsed document to a database or similar situations where you don't need to parse the document just once and save it somewhere. For that, the Parser supports the ability to stringify a parsed AsyncAPI document through the static `AsyncAPIDocument.stringify(...parsedDoc)` method. This method differs from the native `JSON.stringify(...json)` implementation in that every reference that occurs (at least twice throughout the document) is converted into a [JSON Pointer](https://datatracker.ietf.org/doc/html/rfc6901) path.
+
+To parse a stringified document into an AsyncAPIDocument instance you must use the static `AsyncAPIDocument.parse(...stringifiedDoc)` method. It isn't compatible with the native `JSON.parse()` method. 
+
+A few advantages of the solution:
+- string is as small as possible due to the use of [JSON Pointers](https://datatracker.ietf.org/doc/html/rfc6901).
+- all circular references are preserved.
```
[the remark this answers is not in the document]
Comment (alequetzalli) [reply README.md:229]: Just needed some commas and a few grammar fixes :D

[the remark this answers is not in the document]
Comment (magicmatatjahu) [reply README.md]: Thanks! :)

**Comment by magicmatatjahu** [comment]
@jonaslagoni Could you accept again? I had to propagate changes for docs :)

**Review by alequetzalli — APPROVED** [review]
LGTM 🚀

**Review by jonaslagoni — APPROVED** [review]
👍

**Comment by magicmatatjahu** [comment]
@alequetzalli @jonaslagoni Thanks! :)
